pkg/cache/application.go

```diff
@@ -62,6 +63,68 @@ type Application struct {
 	placeholderTimeoutInSec    int64
 	schedulingStyle            string
 	originatingTask            interfaces.ManagedTask // Original Pod which creates the requests
+
+	newTasks *newTasksTracker
+}
+
+// For performance reasons, we track tasks with "New" state for quick retrieval.
+// Application.GetNewTasks (and getTasks() in general) must not be used on critical paths because it
+// uses a read lock inside fsm.FSM which becomes a bottleneck with large number of pods.
+type newTasksTracker struct {
+	sortedTasks *btree.BTree
+	sync.Mutex
```

Comment:
Shouldn't we use an RWMutex here as reads are likely to (significantly) outstrip writes?